src/app/pages/PerpetualPage/components/EditMarginDialog/index.tsx

```diff
@@ -0,0 +1,230 @@
+import { actions } from '../../slice';
+import { isPerpetualTrade, PerpetualPageModals } from '../../types';
+import { TradeDetails } from '../TradeDetails';
+import { AssetValue } from '../../../../components/AssetValue';
+import { AssetValueMode } from '../../../../components/AssetValue/types';
+import {
+  calculateApproxLiquidationPrice,
+  calculateLeverageForMargin,
+  getBase2CollateralFX,
+} from '../../utils/perpUtils';
+import { fromWei, toWei } from 'web3-utils';
+import { usePerpetual_queryPerpParameters } from '../../hooks/usePerpetual_queryPerpParameters';
+import { usePerpetual_queryAmmState } from '../../hooks/usePerpetual_queryAmmState';
+import classNames from 'classnames';
+import { LeverageViewer } from '../LeverageViewer';
+import { toNumberFormat } from '../../../../../utils/display-text/format';
+import { AmountInput } from '../../../../components/Form/AmountInput';
+import { usePerpetual_queryTraderState } from '../../hooks/usePerpetual_queryTraderState';
+import { usePerpetual_accountBalance } from '../../hooks/usePerpetual_accountBalance';
+
+enum EditMarginDialogMode {
+  increase,
+  decrease,
+}
+
+export const EditMarginDialog: React.FC = () => {
+  const dispatch = useDispatch();
+  const { t } = useTranslation();
+  const { pairType: currentPairType, modal, modalOptions } = useSelector(
+    selectPerpetualPage,
+  );
+  const ammState = usePerpetual_queryAmmState();
+  const perpParameters = usePerpetual_queryPerpParameters();
+  const traderState = usePerpetual_queryTraderState();
+  const trade = useMemo(
+    () => (isPerpetualTrade(modalOptions) ? modalOptions : undefined),
+    [modalOptions],
+  );
+  const pair = useMemo(
+    () => trade?.pairType && PerpetualPairDictionary.get(trade.pairType),
+    [trade],
+  );
+  const { available } = usePerpetual_accountBalance(
+    trade?.pairType || currentPairType,
+  );
+
+  const [mode, setMode] = useState(EditMarginDialogMode.increase);
+  const onSelectIncrease = useCallback(
+    () => setMode(EditMarginDialogMode.increase),
+    [],
+  );
+  const onSelectDecrease = useCallback(
+    () => setMode(EditMarginDialogMode.decrease),
+    [],
+  );
+
+  const [margin, setMargin] = useState('0');
+  const [changedTrade, setChangedTrade] = useState(trade);
+
+  const onClose = useCallback(
+    () => dispatch(actions.setModal(PerpetualPageModals.NONE)),
+    [dispatch],
+  );
+
+  const onSubmit = useCallback(
+    () =>
+      // TODO: implement review and excecution for EditMarginDialog
+      dispatch(
+        actions.setModal(PerpetualPageModals.TRADE_REVIEW, changedTrade),
+      ),
+    [dispatch, changedTrade],
+  );
+
+  const [maxAmount, maxAmountWei] = useMemo(() => {
+    if (mode === EditMarginDialogMode.increase) {
+      // Fees don't need to be subtracted, since Collateral is not paid with the Network Token
+      return [Number(fromWei(available)), available];
+    } else {
+      const requiredMargin =
```

Comment:
I would expect this formula to be in some utils (probably not `perpUtils` as those should probably contain just utils provided by Vasili).